engine/crates/federated-graph/src/federated_graph/v2.rs

```diff
@@ -0,0 +1,517 @@
+pub use super::v1::{
+    Definition, EnumId, FieldId, FieldProvides, FieldRequires, FieldSet, FieldSetItem, FieldType, FieldTypeId,
+    InputObjectId, InterfaceField, InterfaceId, Key, ListWrapper, ObjectField, ObjectId, Override, OverrideSource,
+    RootOperationTypes, ScalarId, StringId, Subgraph, SubgraphId, UnionId,
+};
+
+/// A composed federated graph.
+///
+/// ## API contract
+///
+/// Guarantees:
+///
+/// - All the identifiers are correct.
+///
+/// Does not guarantee:
+///
+/// - The ordering of items inside each `Vec`.
+#[derive(serde::Serialize, serde::Deserialize, Clone)]
+pub struct FederatedGraphV2 {
+    pub subgraphs: Vec<Subgraph>,
+
+    pub root_operation_types: RootOperationTypes,
+    pub objects: Vec<Object>,
+    pub object_fields: Vec<ObjectField>,
+
+    pub interfaces: Vec<Interface>,
+    pub interface_fields: Vec<InterfaceField>,
+
+    pub fields: Vec<Field>,
+
+    pub enums: Vec<Enum>,
+    pub unions: Vec<Union>,
+    pub scalars: Vec<Scalar>,
+    pub input_objects: Vec<InputObject>,
+
+    /// All [input value definitions](http://spec.graphql.org/October2021/#InputValueDefinition) in the federated graph. Concretely, these are arguments of output fields, and input object fields.
+    pub input_value_definitions: Vec<InputValueDefinition>,
+
+    /// All the strings in the federated graph, deduplicated.
+    pub strings: Vec<String>,
+
+    /// All the field types in the federated graph, deduplicated.
+    pub field_types: Vec<FieldType>,
+
+    /// All composed directive instances (not definitions) in a federated graph.
+    pub directives: Vec<Directive>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct InputValueDefinition {
+    pub name: StringId,
+    pub field_type_id: FieldTypeId,
+    pub directives: Directives,
+    pub description: Option<StringId>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, PartialOrd)]
+pub enum Value {
+    String(StringId),
+    Int(i64),
+    Float(f64),
+    Boolean(bool),
+    EnumValue(StringId),
+    Object(Box<[(StringId, Value)]>),
+    List(Box<[Value]>),
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, PartialOrd)]
+pub struct Directive {
+    pub name: StringId,
+    pub arguments: Vec<(StringId, Value)>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone)]
+pub struct Enum {
+    pub name: StringId,
+    pub values: Vec<EnumValue>,
```

Comment:
IMHO it's worth changing it immediately to an `EnumValueId` and storing `EnumValue`s in an array, I'll need it to support renaming those efficiently in default values.

Reply:
Ah good point, I agree it's probably a lot less pain in the future, doing that.

Reply:
done in next commit